// commander/index.js
class Command {
  constructor(name = '') {
    this._name = name;
    this._description = '';
    this._version = '';
    this._action = null;
    this.commands = [];
    this.parent = null;
  }

  name(value) { this._name = value; return this; }
  description(value) { this._description = value; return this; }
  version(value) { this._version = value; return this; }
  command(name) {
    const cmd = new Command(name);
    cmd.parent = this;
    this.commands.push(cmd);
    return cmd;
  }
  action(fn) { this._action = fn; return this; }

  outputHelp() {
    const lines = [];
    lines.push(`Usage: ${this._name || 'command'} [options]${this.commands.length ? ' [command]' : ''}`);
    if (this._description) {
      lines.push('');
      lines.push(this._description);
    }
    lines.push('');
    lines.push('Options:');
    lines.push('  -h, --help     display help for command');
    if (this._version) lines.push('  -V, --version  output the version number');
    if (this.commands.length) {
      lines.push('');
      lines.push('Commands:');
      for (const cmd of this.commands) {
        lines.push(`  ${cmd._name.padEnd(12)} ${cmd._description}`.trimEnd());
      }
    }
    process.stdout.write(`${lines.join('\n')}\n`);
  }

  async parseAsync(argv) {
    const args = argv.slice(2);
    if (args.includes('--help') || args.includes('-h') || args.length === 0) {
      this.outputHelp();
      return this;
    }
    if (args.includes('--version') || args.includes('-V')) {
      process.stdout.write(`${this._version}\n`);
      return this;
    }
    const subcommandName = args[0];
    const subcommand = this.commands.find((cmd) => cmd._name === subcommandName);
    if (!subcommand) {
      process.stderr.write(`Unknown command: ${subcommandName}\n`);
      process.exitCode = 1;
      return this;
    }
    if (args.slice(1).includes('--help') || args.slice(1).includes('-h')) {
      subcommand.outputHelp();
      return this;
    }
    if (subcommand._action) {
      await subcommand._action();
    }
    return this;
  }
}

export { Command };
